### ISSUE TITLE
import ConstJson from './config.json' as const;

### ISSUE THREAD

**Issue by slorber** [microsoft/TypeScript#32063]
## Search Terms

json const assertion import

## Suggestion

The ability to get const types from a json configuration file.

IE if the json is:

```json
{
  appLocales: ["FR","BE"]
}
```

I want to import the json and get the type `{appLocales: "FR" | "BE"}` instead of `string`

## Use Cases

Current approach gives a too broad type `string`. I understand it may make sense as a default, but having the possibility to import a narrower type would be helpful: it would permit me to avoid maintaining both a runtime locale list + a union type that contains the values that are already in the list, ensuring my type and my runtime values are in sync.

## Checklist

My suggestion meets these guidelines:

* [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [X] This wouldn't change the runtime behavior of existing JavaScript code
* [X] This could be implemented without emitting different JS based on the types of the expressions
* [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [X] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## Links:

This feature has been mentionned:
- by @qm3ster and @maartenth https://github.com/microsoft/TypeScript/pull/29510#issuecomment-456833107
- by @m-b-davis https://github.com/Microsoft/TypeScript/issues/26552#issuecomment-496925984

**Comment by Hawkbat** [comment]
This would be extremely useful for adding direct support for JSON schemas to TypeScript. This can technically be accomplished with generators right now, but it would be so much more elegant to be able to use mapped types (for example, https://github.com/wix-incubator/as-typed) to map an imported JSON schema to its corresponding TypeScript interface. It isn't currently possible to use this approach with a JSON import since the `type` property of each schema object will be a `string` instead of `'boolean' | 'string' | 'number | ...'`.

**Comment by Porges** [comment]
FWIW I just tried to do this and used the exact same syntax that the issue title uses, if that's any indication of how intuitive it is 😁

**Comment by dontsave** [comment]
I just tried the exact above syntax also. Const assertion is a fantastic tool and it would be incredible to have the ability to assert static json files at import

**Comment by mikeselander** [comment]
I added a note to #26552 and now realize that I put it in the wrong place, so copying it over here :D 

Reading JSON more literally into string types would be a significant improvement to be able to put configs into JSON. 

As an example, the WordPress Gutenberg project is moving towards [a JSON registration schema](https://github.com/WordPress/gutenberg/blob/master/docs/rfc/block-registration.md) for multiple reasons. The list of available category options could and should be tightly limited to the available options. However, due to this bug, we could never enforce a proper category list which effectively breaks TS linting of the file for anyone wanting to use TS when creating Gutenberg blocks or plugins.

**Comment by mscottnelson** [comment]
I've been trying to work on a fix for some of these issues here: https://github.com/pabra/json-literal-typer. If your use-case is relatively straightforward (limited special characters, no escape characters in string literals), then it may satisfy some needs. Would love to have this built-in to the language, but hopefully this will be helpful to some in the interim.

**Comment by Kingwl** [comment]
// CC: @DanielRosenwasser @RyanCavanaugh 

How about syntax `import const ConstJson from './config'` and limited for the json modules.

**Comment by m-b-davis** [comment]
@Kingwl I think this syntax could be slightly more confusing than the alternatives. It could look similar to the import name when viewed in a sequence of imports. It would be good to get some a view on what the preferred syntax would be for everyone.

1 - `import const myJson from './myJson.json';
2 - `const import myJson from './myJson.json';
3 - `import myJson from './myJson.json' as const';

**Personal view:**

#1 - As mentioned above, not sure it's optimal due to distance from import name  
#2 - Perhaps too similar to `const foo = 'abc'`. I think this would at first pass look more like a variable assignment than an import
#3 - This is more similar to the behaviour we have for as const so I would vote for this as the one that fits current design the best.
 
Thoughts? Have I missed any alternative syntax options?

**Comment by m-b-davis** [comment]
Also happy to work on this if it progresses!

**Comment by TheMrZZ** [comment]
I'm for option #3. This one: looks similar to the current "as const" syntax:
```ts
const x = {...} as const
```

It makes it more intuitive. Definitely a killer feature for config-based code if Typescript adopts it.

**Comment by parzhitsky** [comment]
As great as this suggestion is, how should TypeScript interpret the type of property in the original comment?

```json
{
  "appLocales": [ "FR", "BE" ]
}
```

1. Sealed tuplet: `readonly [ "FR", "BE" ]`
1. Tuplet: `[ "FR", "BE" ]`
1. Array of custom strings: `("FR" | "BE")[]`
1. Array of arbitrary strings: `string[]` _<-- the current one_

I think, there's no way for TypeScript to know the desired level of strictness without developer explicitly specifying it somehow, &mdash; and it looks like this will have to be done per each file, rather than once in _tsconfig.json_

**Comment by parzhitsky** [comment]
I think I'm gonna answer my own question 🙂 

The `const` keyword in `as const` is not much about inferring the types literally, as it is about declaring the value as immutable, read-only, sealed. _That in turn_ helps to infer the types more literally, without the worry about being too specific.

With this in mind, it would be intuitive and expected to set the output of *.json modules to be read-only, forbidding any mutable operation on them. This would make it work just like [it currently is working with runtime objects and `as const` assertion](https://www.typescriptlang.org/play?#code/MYewdgzgLgBCBGArGBeGBvAUASAIYAd8AZEYXAGwFMIAuGAbRgCIAxAJSYBpmAhAUSYwAup0wBfGLggxQkKAG5MAeiXYsK7DC0wATpVwATcOQCekwiTJVau-UbCmGrDtyb8mQxRrGYgA).

**Comment by ThomasAribart** [comment]
@parzhitsky I think most would agree on the 1st suggestion, as it is coherent with the present `as const` statement, and as it is the narrowest option (other types can easily be derived from it if needed).

**Comment by pandaniell** [comment]
This is getting even more valuable after Variadic Tuple Types, any thoughts on implementing this yet?

**Comment by mscottnelson** [comment]
For the last half-year or so I have been forcing this behavior by having a somewhat kludgy pre-compile step that reads in a `config.json` like `{"thing":"myVal"}` and exports it as a `config.ts` like `export const Config = {"thing":"myVal"} as const;` and use the resulting type definition on the imported json. (previously I needed to do a lot more, prepending `readonly` everywhere to get the desired array behavior, but at some point that all became unnecessary). It is very helpful during development!

Configuration will likely vary at runtime and thus the content of the json import cannot be known; nevertheless, a `as const` compiled json delivers on all of TypeScript's primary design goals. I can report that having used it to wrangle over-sized configuration json, it has been invaluable in:
- enforcing valid configuration defaults
- inspecting (with "debugger"-like precision) current configuration values while working on code that consumes it
- providing quick semantic help with regards to the shape of configuration subtrees when that shape is not yet defined in a TypeScript definition (ie on the consumption side).

That is to say, from a pragmatic perspective, `import config from './config.json' as const` does most of the things that I find TypeScript most helpful for.

**Comment by lukeapage** [comment]
@RyanCavanaugh you set this as "Awaiting more feedback" - it has 110 thumbs up and a load of comments from people who would find the feature useful. Can it be considered now or at least the tags changed? Or does it require more people to add to the emoji's ?

**Comment by ThomasAribart** [comment]
@RyanCavanaugh This feature would be very helpful for [json-schema-to-ts](https://github.com/ThomasAribart/json-schema-to-ts). You could define and use JSON schemas on one side (API Gateway, swagger... whatever!), and use them in the TS code to infer the type of valid data. Less code duplication, more consistency, everyone would be happier!

**Comment by gabro** [comment]
Chiming in to add another use case: with the new recursive mapped types + string transformations we would like to parse the ICU syntax in translation files in order to extract the params needed to translation strings.

Here's a simplified example:

```typescript
type ParserError<E extends string> = { error: true } & E
type ParseParams<T extends string, Params extends string = never> = string extends T ? ParserError<'T must be a literal type'> : T extends `${infer Prefix}{${infer Param}}${infer Rest}` ? ParseParams<Rest, Params | Param> : Params
type ToObj<P extends string> = { [k in P] : string } 

const en = {
    "Login.welcomeMessage": "Hello {firstName} {lastName}, welcome!"
} as const

declare function formatMessage<K extends keyof typeof en>(key: K, params: ToObj<ParseParams<typeof en[K]>>): string;

formatMessage('Login.welcomeMessage', {firstName: 'foo' })                  // error, lastName is missing
formatMessage('Login.welcomeMessage', {firstName: 'foo', lastName: 'bar' }) // ok
``` 

[Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBACghgJwM4QQUQQg9ggPGqCAD2AgDsATJKJYBASzIHMA+KAXigG9DMcAuKHQCu0AL5QAZFDQBYAFChIsRCngI4AWyS4AKoRLkqNOoyYAaFRu0HSlarQbMOUMhABuqNp0dnbR6n0AfisUdD48AHJ9TWFaKAAjaDgoABt6Ug1UoXAISLZBfWI7YwADABIuRgAzVFgECGr6IjEuSpq69S0xMXayWoQoACUIWjFSqBD1NUQtHRHaSy6bAB8rLQL17QUlaF0sAHkEgCtcGH97EydWFx4AbQBrKEZYAF0oQV9nCQUFAGMsGR4uRbgooOCoAAiAAyWCYjAAdAB3CCpAGaCAAWVGSDgTAgkMEkIAEqjUlhuE1kMAAHJaCASLipOC0OkYsSWFForAYgCEkIUEhZUABQOAv3kFAgf2ZDSg1WEZD+wHogPlOE0cGA2KQuPxuAA0hdjA8ICAsNUcpALYQyCwABSmkCCA2WMCzbSFQ4nM6qCDLHS7G3kO4G14sFgASk+pmYAG4JdUNVqdXqIPbIrD4WRkaj0VicXi8pYuFTWfTBJEk1hIlAxJGFEmEJrtYX8Rms4iufnU0XIiWy7SK1Aq1ga5ZmeWMZWEoha-WFEA])

This currently requires `as const` on the translations object, which we can't do because it lives in a json file.

**Comment by teppeis** [comment]
Related: https://github.com/microsoft/TypeScript/issues/40694 Implement Import Assertions (stage 3)

**Comment by SimonAlling** [comment]
I also found my way here because I ran into type errors when trying to use a JSON module in a somewhat strongly typed context. In addition to the already mentioned use cases, being able to import JSON modules "`as const`" would allow one to do this:

```tsx
import { AsyncApiInterface } from "@asyncapi/react-component"
import React from "react"

import schema1 from "./schema1.json"
import schema2 from "./schema2.json"

function dropdown(schemas: readonly AsyncApiInterface[]) {
  return (
    <select>
      {schemas.map(schema => <option>{schema.info.title}</option>)}
    </select>
  )
}

dropdown([ schema1, schema2 ]) // type error at the time of writing

```

**Comment by Tommos0** [comment]
Possible workaround (if your file is called `petstore.ts`):
`echo -E "export default $(cat petstore.json)" as const > petstore.json.d.ts`

Now import as normal to get the literal type.

**Comment by apancutt** [comment]
> Possible workaround (if your file is called `petstore.json`):
> `echo -E "export default $(cat petstore.json) as const" > petstore.json.d.ts`
> 
> Now import as normal to get the literal type.

To avoid _"The expression of an export assignment must be an identifier or qualified name in an ambient context."_ (TS2714) errors, use:

`echo -E "declare const schema: $(cat petstore.json); export default schema;" > petstore.json.d.ts`

**Comment by ImRodry** [comment]
Hope the issue above gives some more information that could help implement the feature suggested in this one.

**Comment by parzhitsky** [comment]
There is a stage-3 ECMAScript proposal called "JSON modules" (see, https://github.com/tc39/proposal-json-modules), derived from "import assertions" (see, https://github.com/tc39/proposal-import-assertions), yet another stage-3 proposal:

```
import json from "./foo.json" assert { type: "json" };
```

I know that "type" here means something completely different than what it means in TypeScript, and also we might not need to actually _assert_ anything (just infer typings), – but this still feels like something related, maybe we should piggyback on this one.

**Comment by askirmas** [comment]
Meanwhile

#### `json-d-ts.sh`
```bash
#!/bin/bash
PATTERN="*.schema.*json"
DECL_POSTFIX=".d.ts"

find . -type f -iname "$PATTERN$DECL_POSTFIX" -exec rm {} \;
JSONS=($(find . -type f -iname "$PATTERN"))

for file in "${JSONS[@]}"
do
  git check-ignore --quiet "$file" ||\
  printf "/** Generated with \`./json-d-ts.sh\` */\ndeclare const data: $(cat $file)\nexport = data" > "$file.d.ts"
done
```

#### `package.json`
```json
{
  "scripts": {
    "preparation": "./json-d-ts.sh"
  }
}
```

**Comment by DetachHead** [comment]
my use case is a json schema that gets converted to type definitions using a utility type:
```ts
type Foo = FromJsonSchema<{
    type: 'object'
    properties: {
        foo: {
            type: 'object'
            properties: {
                asdf: {
                    type: 'string'
                }
            }
            additionalProperties: false
        }
        bar: {
            type: 'array'
            items: {
                type: 'string'
            }
        }
    }
    additionalProperties: false
}>

// Type '{}' is missing the following properties from type '{ foo: { asdf: string; }; bar: string[]; }': foo, bar
const foo: Foo = {}
```
currently the schema has to be defined inline, which isn't ideal. would be nice if i could do something like
```ts
import const schema from './schema.json'

type Foo = FromJsonSchema<typeof schema>
```

**Comment by flixcor** [comment]
I made a bundler plugin (webpack/rollup/vite) that helps with this:
https://www.npmjs.com/package/unplugin-json-dts
https://github.com/flixcor/unplugin-json-dts/

**Comment by agrajak** [comment]
for those who use @askirmas code and want to escape `\n` string in JSON, use this one
```bash
#!/bin/bash
# https://github.com/microsoft/TypeScript/issues/32063#issuecomment-916071942
PATTERN="*.schema.*json"
DECL_POSTFIX=".d.ts"

find . -type f -iname "$PATTERN$DECL_POSTFIX" -exec rm {} \;
JSONS=($(find . -type f -iname "$PATTERN"))


for file in "${JSONS[@]}"
do
  text=$(cat $file | jq -sR .)
  length=${#text}-2
  git check-ignore --quiet "$file" ||\
  printf "/** Generated with \`./json-d-ts.sh\` */\n/* eslint-disable */\ndeclare const data: $(echo ${text:1:$length})\nexport = data" > "$file.d.ts"
done
```

**Comment by Swazimodo** [comment]
What needs to be done to move this forward? From the thread it looks like `import myJson from './myJson.json' as const` is the agreed upon syntax.

**Comment by thw0rted** [comment]
I don't see any activity from the TS team, which is surprising given that it has nearly 300 thumbs-up reactions.  @RyanCavanaugh marked it "awaiting feedback" but there appears to be plenty of that.  I think all the community can do now is wait?

**Comment by albnnc** [comment]
Just casually found this issue in a process of understanding that const-typed json imports are currently impossible in TS. I also think that this feature would be incredibly useful for DX, which is the primary target for TS as far as I can understand. Would like to see this implemented.

**Comment by lukeapage** [comment]
300 isn’t that many - sort open issues by thumbs up and this comes on the 2nd page. A lot that are ahead of this really surprise me. More thumbs up might help, the teas has said earlier they look through issues sorted by it periodically.

**Comment by willmartian** [comment]
Suppose I wanted to import a JSON for the _sole purpose_ of using it as a const type. **This syntax hasn't been suggested yet**, but it feels consistent to me:

`import type MyElement from "custom-element-manifest.json"`
    
(As that code snippet alludes to, a cool use case for this would be mapping types from things like the [custom elements manfiest](https://github.com/webcomponents/custom-elements-manifest).)

**Comment by btoo** [comment]
maybe `import type` can be used for both normal and `const` json `import`s:

---
```typescript
import type SomeJsonType from "some-json.json"
```
is equivalent to
```typescript
import someJson from "some-json.json"
type SomeJsonType = typeof someJson
```
---
```typescript
import type SomeConstJsonType from "some-json.json" as const
```
is equivalent to
```typescript
import someConstJson from "some-json.json" as const
type SomeConstJsonType = typeof someConstJson
```

**Comment by Platane** [comment]
The `import type` syntax is fully erased by typescript. Which ensure that the module is not require at runtime.

Whereas with 
```ts
import someConstJson from "some-json.json" as const
type SomeConstJsonType = typeof someConstJson
```

You might end up with the content of `some-json.json` in your bundle ( depending on how good your bundler is at treeshaking ). 

It's not strictly equivalent.

**Comment by somebody1234** [comment]
Since there's been no maintainer activity I might as well give a shot at implementing this.
The parser changes seem trivial enough, the typechecker changes are a little harder. For context here's the code determining whether an expression is `as const`:
```ts
        function isConstContext(node: Expression): boolean {
            const parent = node.parent;
            return isAssertionExpression(parent) && isConstTypeReference(parent.type) ||
                isJSDocTypeAssertion(parent) && isConstTypeReference(getJSDocTypeAssertionType(parent)) ||
                (isParenthesizedExpression(parent) || isArrayLiteralExpression(parent) || isSpreadElement(parent)) && isConstContext(parent) ||
                (isPropertyAssignment(parent) || isShorthandPropertyAssignment(parent) || isTemplateSpan(parent)) && isConstContext(parent.parent);
        }
```
As you can see, it's determined completely by parent nodes... which makes sense since the only source of `as const` is, well... `as const`.
So the main problem with this change is having to associate multiple types to a single node.
(Another problem would be figuring out how to propagate "I want the const type of this" through the module resolution + typechecking)
(If anyone has any ideas, please feel free to comment)

**Comment by bombillazo** [comment]
It would be phenomenal to have this feature!

**Comment by lobotomoe** [comment]
"Awaiting More Feedback"? Srsly? :)

**Comment by slifty** [comment]
I made the same mistake as someone else and commented in #26552 but should post here!  Hopefully this isn't adding to noise...

---

Wanted to chime in here to agree with the limitations of the current implementation and provide another use case.

I'm using [ajv](https://ajv.js.org/guide/typescript.html) to create type guards for my types that are powered by a Type-validated-JSONschema.  This is really handy because it's possible for ajv to throw type errors if the schema (and therefore the type guard) does not match the Type definition.

This all works fine if defined inline:

```
export interface Foo {
  bar: number;
};

const validatedFooSchema: JSONSchemaType<Foo> = {
  type: 'object',
  properties: {
    bar: {
      type: 'integer'
    },
  },
  required: ['bar']
};
export const isFoo = ajv.compile(validatedFooSchema);
```


However, I want to define my json schema in an external json file so that I can reference it in other places (e.g. in an OpenAPI spec).  **I *cannot* do this with the current TypeScript json import implementation**, since this schema is not interpreted as the required string literals but as generic `string` types:

**fooSchema.json**
```
{
  "type": "object",
  "properties": {
    "bar": {
      "type": "integer"
    },
  },
  "required": ['bar']
}
```

**Foo.ts**
```
import fooSchema from '../fooSchema.json';

export interface Foo {
  bar: number;
};

const validatedFooSchema: JSONSchemaType<Foo> = fooSchema;
export const isFoo = ajv.compile(validatedFooSchema);
```

^ `validatedFooSchema` errors with:

```
Types of property 'type' are incompatible.
        Type 'string' is not assignable to type '"object"'
```

**Comment by thw0rted** [comment]
@slifty  you might be interested in https://github.com/vega/ts-json-schema-generator -- I had a similar problem, and used that to go from TS interface(s) -> JSON schema, rather than the other way around.

**Comment by lbguilherme** [comment]
This would be really useful for the [`as-typed`](https://www.npmjs.com/package/as-typed) package, as it would allow to import and use types from JSON schemas directly.

```ts
import type { AsTyped } from "as-typed";
import type schema from "./schema-data.json" as const; // <--

export type Business = AsTyped<typeof schema["components"]["schemas"]["Business"]>;
```

Currently the only way to use it is to copy the JSON file into a TypeScript file with `as const` added.

**Comment by varanauskas** [comment]
On the same note as @lbguilherme this would also be useful for a [library I am creating](https://github.com/varanauskas/oatx)
by allowing users to import OpenAPI spec straight from JSON without having to do the 
```ts
declare const data: {};
default export data;
```
song and dance

**Comment by mogafk** [comment]
it's very useful feature

**Comment by Ustice** [comment]
I have had two uses for this on my current big project. The first is loading static data. I would much rather just define it, and infer the types from there. It really stinks to have to do that in two places.

The second is that we are making a utility to manage environment variables. I would really like to have all of the used environment variables in a json file that I can load, and turn into a tuple of static values. When someone references a new one in the code, it will update the json file with the new key. I can totally do this by creating a code generator, but it would be really nice if I could just statically import the data `as const`.

**Comment by adlerfaulkner** [comment]
Would also be extremely useful in [Standard SDK](https://github.com/comake/standard-sdk-js), which dynamically reads OpenAPI specs from JSON to automagically create an SDK with any REST api. 

I'm getting:
```
Type 'string' is not assignable to type '"path" | "header" | "query" | "cookie"'
```
when an imported OpenAPI spec is used as an argument which is typed with an OpenAPI validation type to a function that creates the SDK. I want to use the value of this argument (the OpenAPI spec) to add autocompletion to the SDK it creates. It is critical to our project to allow users to import OpenAPI specs "as is" and not have to got through the process of converting them to javascript objects first.

**Comment by KholdStare** [comment]
@RyanCavanaugh Just another friendly poke :) It has been more than 3 years (!!!) since this feature was requested. There are quite a few projects that would greatly benefit from this.

**Comment by DetachHead** [comment]
i thought #51865 would've allowed for a workaround for this, but i guess not
```jsonc
// foo.json
{
    "foo": "bar"
}
```
```ts
declare const narrow: <const T>(value: T) => T
const narrowed = narrow((await import('./foo.json')).default) // not narrowed. type is still {foo: string}

**Comment by somebody1234** [comment]
yeah unfortunately not, because `import` is the thing resolving the type

**Comment by ScreamZ** [comment]
Is there any way to achieve this for now?
I'm trying to introspect a `model.json` file from prismic to generate some custom types.

**Comment by TSMMark** [comment]
In case this helps anyone, I'm using [babel plugin codegen](https://www.npmjs.com/package/babel-plugin-codegen) to generate types by reading the json files from disk. Very cumbersome, but better than writing types manually.

**Comment by j-murata** [comment]
Is there any chance that the [`const` type Parameters in TS 5.0](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/#const-type-parameters) will solve this problem?

**Comment by DetachHead** [comment]
@j-murata nope. see these two comments above https://github.com/microsoft/TypeScript/issues/32063#issuecomment-1409510267 https://github.com/microsoft/TypeScript/issues/32063#issuecomment-1409556488

**Comment by PineSongCN** [comment]
> @RyanCavanaugh you set this as "Awaiting more feedback" - it has 110 thumbs up and a load of comments from people who would find the feature useful. Can it be considered now or at least the tags changed? Or does it require more people to add to the emoji's ?

Four years have passed.

**Comment by roninjin10** [comment]
Would be extremely useful.   Without this we need to write a cli tool that generates a .ts file from a json so I can as const it.

**Comment by matthew-dean** [comment]
Just to comment on this thread, I think importing `as const` for a JSON is a bit of the wrong direction, since that doesn't really successfully coerce all the types that you might expect into the shape you actually want.

A surprising behavior of `satisfies` is that it does type coercion, and not just type-checking.

For instance, you may not realize it (or at least I didn't), but these two objects will have two different types:

```ts
let foo = {
  variant: 'primary'
}
let bar = {
  variant: 'primary'
} satisfies { variant: 'primary' }
```
The first object's `variant` key has a string type. The second object's `variant` key has a string literal type where `primary` is the only valid value.

Meaning that following the above with this will cause an error:
```ts
bar = foo
```

So, rather than 
```ts
import myJson from './myJson.json' as const
```
I think the syntax that actually has the flexibility to solve different problems in this thread would be more like:
```ts
import myJson from './myJson.json' with { satisfies: SomeType }
```
That would / could coerce a type into string or number literals exactly where you want it, and preserve the existing type inference where you didn't.

**Comment by RyanCavanaugh** [comment]
Sidenote: Please stop noting the linear passage of time.

**Comment by somebody1234** [comment]
tl;dr: "please stop noting the linear passage of time" is the opposite of constructive feedback.


however, they are expected to be responded to in a timely fashion, rather than having no official feedback whatsoever. and it's not even like it'd be that difficult to constructively give feedback:

- this is not being considered for &lt;x&gt; reason
- this is more trouble than it's worth because &lt;y&gt; makes it hard
- we are accepting PRs, &lt;this&gt; is how it should be done if anyone's interested
- we are still awaiting feedback, &lt;z&gt; is still unclear
- there are issues with it, for example &lt;a&gt;, &lt;b&gt; and &lt;c&gt;

i personally think it's a little silly to address the two comments that note the passage of time and ***completely ignore*** the other 95%.

it's also worth noting that saying "please don't note the linear passage of time" will not stop people from saying that

**Comment by RyanCavanaugh** [comment]
There are thousands of open suggestions, and I am doing the best I can to leave comments explaining why certain features haven't made the iteration plans, as well as explicitly rejecting certain features that have been opened for a long time. You can see this in places like minification, regex types, typed exceptions, etc.. I'm doing the best I can and I'm asking people to just act in a way that makes it easier for me to do that, which I think is aligned with everyone's interests since posting the comments about how long something has been open doesn't help anyone.

Part of the difficulty in writing those comments is honestly the sheer volume of nonspecific feedback ("Any updates?" "This is has been open for N years") that doesn't contribute meaningfully to the discussion, and leads people to make duplicative comments because things that have already been said are hiding behind "Load [n] more comments" blocks.

**Comment by somebody1234** [comment]
tl;dr: i don't think there's a "sheer volume" of nonspecific feedback here - *at the very least* it doesn't make things "difficult" in this thread specifically. as i have already pointed out there are only two topics about the amount of time passed. there are (currently) 40 hidden items. even *completely removing* said comments would reduce the amount of messages behind "load [n] more comments" blocks by 5% which is an insignificant fraction.

sure, that's fair. regardless:
- i'm still not convinced that your first comment is constructive.
- yes there are thousands of open suggestions but hear me out: this is the *tenth most upvoted suggestion*.
  - *please* also note that i believe this is is not:
    - a major undertaking that creates additional complexity for the end-user, as with regex types
    - not feasible/not that useful without major DT changes, as with typed exceptions
    - out of scope (or already handled by dedicated build tools for years), as with minification
- as i've already pointed out, said non-constructive comments are a vanishingly small fraction of this thread. i think it'd be much faster, less controversial, and just better in general for everyone involved to just silently mark them as off-topic instead of intentionally bringing attention to them.

**Comment by miguel-leon** [comment]
bro are you paying him? why u demanding so much?

**Comment by sebastian-fredriksson-bernholtz** [comment]
@matthew-dean I think you're missing a few very important points. Most importantly, `satisfies` and `as const` fulfil very different purposes. The purpose of `as const` is to assert (cast/coerce) the type of a literal value into the narrowest possible type. `satisfies` is for validating that the type of a value is a subtype of another type **without** making any type assertion (cast/coercion)!

Lets look at `satisfies` in some detail: 

1. It's better to discuss this issue in terms of 'type narrowing" and "type widening" rather than the much broader "type coercion".
2. `satisfies` does not technically do any type coercion - or type narrowing/widening for that matter. The entire purpose of the `satisfies` operator is to verify that the type of a value conforms - is a subtype - to another wider type without widening the type of the value . It most definitely is not intended to do type narrowing!

So if `satisfies` doesn't do type narrowing, what is going on in your example?

1. A literal value in TypeScript - unlike variables, etc - isn't really of a single **specific** TypeScript type. While it is of a specific runtime type - and the runtime type corresponds to a TypeScript type - the value can be interpreted as either the primitive type or the literal type. 
Eg. The literal string `"hello"` is of runtime type `string`, but can be interpreted as both the TypeScript primitive type `string` and the string literal type `"hello"`.
2. When a variable is assigned to without a type annotation, TypeScript infers that the variable is of the same type as the type of what it's being assigned. Without any further information TypeScript will interpret all literal values as their primitive types rather than their literal types.
3. However, if a literal value `satisfies` a literal type, Typescript can only interpret it as the literal type, since interpreting it as the primitive type would fail the type verification.

So now we can explain your example despite `satisfies` not technically doing any type casting:
```typescript
let foo = { variant: "primary" } 
// can be interpreted as either primitive or literal type, is interpreted as primitive

let bar = { variant: "primary" } satisfies { variant: "primary" } 
// invalid if interpreted as primitive  - 'string' not assignable to '"primary"' - is interpreted as literal
```

I'd be interested in knowing how you think `as const` 
>  doesn't really successfully coerce all the types that you might expect into the shape you actually want.

The only thing that I could think of is that you don't want it to be readonly. However, with the proposed `as const` import and the existing `satisfies` operator you could do whatever you want with the structure:

```typescript
const foo = {
  variant: "primary",
  invariant: "green"
} as const
// equivalent of importing as const, type is { readonly variant: "primary"; readonly invariant: "green"; }

const bar = {
  ...foo,
  variant: "secondary",
  additional: "stuff"
} satisfies {invariant: "green", variant: string, additional: "stuff"}
// bar is of type  { variant: string; invariant: "green"; additional: "stuff" }
```

If you're talking about something similar to what the original commenter had in mind it can also be achieved through derivation as @parzhitsky [mentioned](https://github.com/microsoft/TypeScript/issues/32063#issuecomment-682173707):
```typescript
const foo = {
  appLocales: ["FR","BE"]
} as const
// type { readonly appLocales: readonly ["FR", "BE"]; }

type Locale =typeof foo["appLocales"][number]
// type "FR" | "BE"

const locale: Locale = "FR"
```

**Comment by matthew-dean** [comment]
@sebastian-fredriksson-bernholtz 

I would buy your argument that `satisfies` isn't doing any type coercion except for two points:

```ts
let foo = 'one' satisfies 'one' | 'two'
```
In the above example, `foo` does satisfy the type given, yet the type of `foo` is still `string`. If `one` is a value within an object, the behavior flip-flops, and type coercion occurs.

In the TypeScript documentation, it explicitly says this:

> The new satisfies operator lets us validate that the type of an expression matches some type, without changing the resulting type of that expression.

Using the `satisfies` operator changes the resulting type of the expression (sometimes). The documentation is false, or the behavior is a bug.

That said, I filed [this issue](https://github.com/microsoft/TypeScript/issues/54488) which has been kept open as a separate problem / solution, so maybe one of these ideas will prevail.

Personally, I don't like using `satisfies` changes the type outcome, contrary to the documentation, so I dunno, maybe `as const` is better and `satisfies` should be fixed. 🤷‍♂️

**Comment by sebastian-fredriksson-bernholtz** [comment]
@matthew-dean 

There is still **no "type coercion"**, only type inference. And while my explanation isn't entirely correct, it becomes so with a slight change of wording from 
> can only interpret it as the literal type

to

>  is "better" to interpret it as the literal type

So, what is going on in your example?

TypeScript is doing it's best trying to infer what type `foo` should be from the information you're providing and being as unobtrusive as possible. Since `foo` is a `let` - as opposed to `const` - it can infer that you intend `foo` to change. Hence, it's "better" to interpret `'one'` as `string` than as `'one'`. 

I do think it's bug and worth an issue to request that `foo` is inferred to be `'one' | 'two'` rather than `string` in your case, but it definitely shouldn't be inferred as "`one`".

The documentation could be clearer, but if you think of literal values as not having a specific type, the documentation isn't incorrect. `satisfies` does not change the type of the expression, the expression is not of a (specific) type.  

As you can see from this simple example `satisfies` has no impact on the inferred type if the value "is of a specific type":
```typescript
let foo = 'one' as 'one' satisfies 'one' | 'two'
// foo is type 'one' because 'one' is of type 'one'
```

**Comment by matthew-dean** [comment]
> satisfies does not change the type of the expression, the expression is not of a (specific) type. satisfies simply provides context for how to infer the type when the type is yet to be determined.

To me, that's a distinction without a difference. "without changing the resulting type of that expression" is the key phrase in the documentation e.g. the resulting type should be identical whether `satisfies` is present or not. Calling it "inference" vs coercion is also just semantics in this instance. It _would_ have been one type, but it is _not_, _because_ of the `satisfies` keyword.

To be clear, I'm fine if `satisfies` has this behavior _if_ it is consistent and _if_ it's noted in the documentation that `satisfies` does have an influence on the "resulting type". That's fair, isn't it?

As to this:

> I'm still interested in what you think is the problem with as const assertion?

There were a few comments in this thread about how exactly `as const` should behave. I think it really depends on how TypeScript restricts how the inferred type can be used.

I guess if it imports JSON as the narrowest type possible, that's fine, but even in testing `as const` in the TypeScript playground, it doesn't infer types in the way I expect.

Take this for example:
```ts
const foo = [
  {
    variant: 'primary',
    num: 1
  },
  {
    variant: 'secondary',
    num: 2
  }
] as const
```
This results in this type:
```ts
const foo: readonly [{
    readonly variant: "primary";
    readonly num: 1;
}, {
    readonly variant: "secondary";
    readonly num: 2;
}]
```
But what I really wanted was:
```ts
type Foo = Array<{
  variant: 'primary' | 'secondary'
  num: number
}>
```
or even
```ts
type Foo = Array<{
  variant: 'primary' | 'secondary'
  num: 1 | 2
}>
```
There's no real way to tell `as const` what to do. Within the code, as we've been discussing, the only way to coerce or "help infer", whatever you want to call it, along with type-check the object is with `satisfies`.

If I do:
```ts
const foo = [
  {
    variant: 'primary',
    num: 1
  },
  {
    variant: 'secondary',
    num: 2
  }
] satisfies Array<{ variant: 'primary' | 'secondary', num: number }>
```
.....well, actually we still don't quite get there, because it still doesn't infer the type that we explicitly gave it, even though it correctly satisfies it. (Honestly, the more I experiment with `satisfies` in relation to this and related issues, the more I'm confused with how it works or is supposed to work.)

We get a type that is:
```ts
const foo: ({
    variant: "primary";
    num: number;
} | {
    variant: "secondary";
    num: number;
})[]
```
...which is clunky and not what I meant for it to infer, but okay. (And doesn't infer the type similarly to the non-array example, so I'm just more confused.) But still, the point is that I can nudge the TypeScript inference using `satisfies` in a way I cannot using `as const`. And, more importantly, I can type-check using `satisfies` while I can't with `as const`.

I think though, to be fair, using `as const` might be "good enough" for most scenarios? And maybe could be further inferred or coerced using some kind of advanced type construct. I'm not sure I would put `as const` on the `import` syntax, instead of extending the new `with` syntax for imports, but maybe it's fine? 🤷‍♂️

**Comment by osdiab** [comment]
This is all stimulating and nice but seems pretty unrelated to the original issue and you’re buzzing subscribers emails without actually indicating progress towards the original goal. Maybe better to have that discussion in a separate GitHub issue/discussion if the above is very important to you?

**Comment by roninjin10** [comment]
Resolving all jsonMOdules as const is dangerous for performance IMO.  It's easy for one engineer to turn that on while another one adds an extremely large json array that requires type checking linearly everytime it's used

**Comment by matthew-dean** [comment]
@roninjin10 I don't see how. The JSON file would still be inferring X type based on Y value. So the actual process involved would be the same. In fact, one could argue that it should take less time because with some values, it can directly assign the type without widening, but that's entirely dependant on TS internals.

Related: has the argument been made in this thread that `as const` should maybe be the default type for JSON, such that `as const` shouldn't be needed? I assume so, but maybe I missed it. I'm wondering what the downside would be for importing as a constant value. If you think about it, string data from a file should already be immutable. I guess there would be a rare risk of breaking code somewhere though.

**Comment by Peeja** [comment]
@matthew-dean Yes, to _infer_ the type, but once it's inferred, if it's a gigantic union of string literals instead of `string` (I believe) it would be expensive to throw around the code _after_ it's been inferred.

OTOH, if it's not expensive or otherwise concerning, it seems to me like it would be the ideal default.

**Comment by toomuchdesign** [comment]
Stumbled upon this thread in the context of generating TS JSON schemas (type-inferrable with [json-schema-to-ts](https://github.com/ThomasAribart/json-schema-to-ts)) from openAPI.

If we considered JSON schema as an "interchange format" able to express both data type and runtime validation, then why not making JSON schema a first class citizen of TS itself? This would mean providing a native way to infer types from any valid JSON schema. I know it's a lot of stuff :) 

In case someone is struggling with openAPI -> TS JSON schema conversion, I've put together this NPM package to try and automate this flow: https://github.com/toomuchdesign/openapi-ts-json-schema

**Comment by slorber** [comment]
Since [ES Import Attributes (Stage 3)](https://github.com/tc39/proposal-import-attributes) are going to [be worked on for TypeScript 5.3](https://github.com/microsoft/TypeScript/issues/55486), I was wondering if it's not a viable option to import JSON as const?

```ts
import data from "./data.json" with { type: "json-const" };
```

I'm not sure about what to use for the `type` though.

**Comment by Malix-off** [comment]
Would const assertion after declaration be considered too?

**Comment by MatthD** [comment]
Hello everyone, so ATM what is the right solution (if it exist) to match json import with some literals and have the codebase happy ?

**Comment by slorber** [comment]
Not a drop-in workaround for json const, but you can type a json file manually.

@mattpocock has a little article on it: https://www.totaltypescript.com/override-the-type-of-a-json-file

<img width="801" alt="CleanShot 2023-12-08 at 13 57 42@2x" src="https://github.com/microsoft/TypeScript/assets/749374/5cfd2da7-3c5b-49ff-bfc3-e9630c98160a">

**Comment by shamrin** [comment]
@slorber Unfortunately `.d.json.ts` definition makes TypeScript ignore the *content* of JSON file. Now json file can have a shape that does not match its type and one would never notice.

**Comment by CodeFromAnywhere** [comment]
Hi I found this issue since I think it would be a great addition to infer Types from Json Schemas! See https://github.com/ThomasAribart/json-schema-to-ts/issues/200

**Comment by couhajjou-apps** [comment]
as a workaround, should we implement a simple converter json2js, it would take a my.json file an create a my.js file.

//my.json
{
  "hello": "string"
}

//my.js
export const my =  // <- add this line
{
  "hello": "string"
}

all it needs to do is to add the first line

and then we can 
import { my } from 'my.js'

**Comment by GabenGar** [comment]
What do you do when the json filename is an invalid symbol name, i.e. it has hyphens?

**Comment by SimonCockx** [comment]
@couhajjou-apps @GabenGar 
Don't reinvent the wheel:
https://quicktype.io/typescript

**Comment by parzhitsky** [comment]
@couhajjou-apps If we're going that route, I'd rather use default export rather than the named one to avoid dependending on the arbitrary filename:

```diff
- {
+ export default {
    "foo": 42,
    "bar": 17
  }
```

**Comment by couhajjou** [comment]
> [@couhajjou-apps](https://github.com/couhajjou-apps) [@GabenGar](https://github.com/GabenGar) No need to reinvent it:
> 
> * As "regular" interface: https://quicktype.io/typescript
> * As const interface (what this issue is about): https://www.npmjs.com/package/ts-json-as-const

I don't see how it's related and why it has 3 thumbs up. Weird.

**Comment by couhajjou** [comment]
> [@couhajjou-apps](https://github.com/couhajjou-apps) If we're going that route, I'd rather use default export rather than the named one to avoid dependending on the arbitrary filename:
> 
> - {
> + export default {
>     "foo": 42,
>     "bar": 17
>   }

Good idea !

**Comment by BinToss** [comment]
> What do you do when the json filename is an invalid symbol name, i.e. it has hyphens?


~~I'm unsure of what you're asking.~~
Oh! The symbol never needs to be the same as the imported module's filename. It can be arbitrary e.g.
```js
Import hyphenJson from './-.json' with { type: 'json' }
```

This discussion isn't about importing JSON modules nor is it about importing them with an attribute so it's imported only with a JSON parser (that was resolved by  https://github.com/tc39/proposal-json-modules).

This discussion is about getting the type information from the imported object as if it was an object defined with the TypeScript syntax `as const`, though what they want differe from how `as const` actually works.

```ts
const arr0 = [
  "entry0",
  1,
];
// typeof arr === (string | number)[]

const arr1 = [
  "entry0",
  1,
] as const;
// typeof arr1 === [ "entry0", 1 ]
```

If they want `["entry0" | 1]` or something similar, they'll have to use type manipulation like `DeepWritable<T>` from the ts-essentials package.

**Comment by mscottnelson** [comment]
The exact thing that `as const` does works extremely well for the config use case as long as you invert the relationship to basically say that the JSON `satisfies` the type of the places where it is used. However, in 2024, there's a lot more tooling around that makes it much easier to do runtime validation of ts types, which ultimately is the more correct approach anyway.

There are still some annoying type narrowing / widening problems that can arise, but `satisfies` can solve many of them without overly verbose definitions.

**Comment by keithriver** [comment]
Having stricter const types for json import would be pretty useful for developing vscode extensions so package.json could be used for source of truth for all the view/command names.

I was honestly surprised that it's not using narrowest types for strings as the content of strings in a json file is pretty much constant on import.

**Comment by djibarian** [comment]
Please tell me how to type `appLocales` as `'FR' | 'BE'` instead of `string` from the imported JSON, because after two days I can't figure it out. Any solution or workaround, ugly or not, which doesn't imply duplicating `'FR' | 'BE'` outside the JSON or having to generate a `.d.ts`.

**Comment by abumalick** [comment]
> Please tell me how to type `appLocales` as `'FR' | 'BE'` instead of `string` from the imported JSON, because after two days I can't figure it out. Any solution or workaround, ugly or not, which doesn't imply duplicating `'FR' | 'BE'` outside the JSON or having to generate a `.d.ts`.

Best way for me is transforming to a ts file and exporting the object as const

**Comment by djibarian** [comment]
And what if you can't transform the file to a ts because is shared with the backend, which uses a different language?

**Comment by ClaudiuCeia** [comment]
> The declaration file is emitted during a build and continuously based on a watcher when the dev server of the project runs ([ref](https://v4.next-intl.dev/docs/workflows/typescript#messages-arguments)).

That's a great "good enough" solution to be honest @amannn 

It sadly doesn't work for my use case where I'm importing a JSON from URL ([Deno](https://github.com/denoland/deno/issues/16583)). I'm also curious what would it take to move this forward.

**Comment by GabenGar** [comment]
I am not sure if there can be anything forward there. Doing network requests in the middle of _static_ type analysis is way outside of the scope, even for Microsoft. Typescript plugins already have problems of `tsc` and whatever type check command used by a framework of that plugin not agreeing on the final types. Adding an even more frail source of type mismatches definitely won't make it easier.

**Comment by phaux** [comment]
@GabenGar Deno's fork of TypeScript already does the fetching. It just doesn't get the const type.

**Comment by ClaudiuCeia** [comment]
@GabenGar There's no network IO involved during static type analysis with Deno. Imports are fetched and cached locally just like you'd install dependencies from npm. `tsc` still runs it's type checks on local files. See [the documentation](https://docs.deno.com/runtime/fundamentals/modules/#https-imports) for more details, this is not Deno reinventing the wheel by the way, Go allows http imports too.

Now, as for Microsoft not being able to pull this off, you'd be surprised, take a look at [F# type providers](https://learn.microsoft.com/en-us/dotnet/fsharp/tutorials/type-providers/), particularly [FSharp.Data](https://fsprojects.github.io/FSharp.Data/) which is something I'd love to see in Typescript as well.

**Comment by GabenGar** [comment]
@ClaudiuCeia 
> There's no network IO involved during static type analysis with Deno.

How does it fetch over a network module without a network IO then?

>` tsc` still runs it's type checks on local files.

You are asking to do that on network. With built-in typescript module guessing logic, which naturally results in a request spam.

> Now, as for Microsoft not being able to pull this off, you'd be surprised

On the contrary, typescript being a package which doesn't require some sort of subscription or at least 20 GB of dependencies is very unusual for a Microsoft product. I am absolutely not surprised their own language does casually make http requests at any point of its development flow.  I bet debugging dependency problems arisen from https errors is super fun too.

**Comment by ClaudiuCeia** [comment]
@GabenGar I guess you haven't read the documentation or understand module resolution in Deno

> How does it fetch over a network module without a network IO then?

It doesn't. The dependency is being fetched and cached before static type checks, the same way you `npm install` and fetch from a remote repository. `tsc` is not running during your `npm install` why would it do so in Deno? Why would you even assume http requests are involved at compile time? You seem convinced that that's what happens, we're talking about an open-source project so feel free to just reply with a link to the code if that's your stance. 

> With built-in typescript module guessing logic, which naturally results in a request spam.

Module guessing logic is exactly what Deno is avoiding but it feels pointless to argue the point at this stage. 

---

The on topic point remains that there's interest in this functionality from the Deno side as well, and I'm still looking to figure out what the status for this is or what would it take to push `as const` json imports further. With a disclaimer that having support for this in Deno doesn't change the way this would work in any way, in case that's not clear.

**Comment by JameEnder** [comment]
Would love to see this being implemented; the workaround mentioned by @amannn works, but adds more "pointless" complexity and tools to maintain, especially if you don't use Next.js and have to write it yourself.

**Comment by Tom-de-Boer** [comment]
I could also very much use this. I am trying to create types from a json-schema.

**Comment by thw0rted** [comment]
@Tom-de-Boer , if it helps at all, our approach has been to create a type decorated using `class-validator`, then run that class through `class-validator-jsonschema`.  The type declarations (`.d.ts` file) can be shared between projects to provide types to your IDE, the decorated class itself can be used for runtime validation via `class-validator` (has convenient hooks already in NestJS), and the compiled JSON schema can be distributed and JSON imports validated with `ajv` or whatever.  Of course it requires some extra infrastructure to build and distribute the various formats of schema but once those are in place, you can capture your type in one place and check validity anywhere.

**Comment by TechQuery** [comment]
> the decorated class itself can be used for runtime validation via `class-validator` (has convenient hooks already in NestJS), and the compiled JSON schema can be distributed and JSON imports validated with `ajv` or whatever.

@thw0rted You can validate data with decorated classes in front-end directly, just do it now with Decorator stage-2, or wait for new Parameter Decorator proposal stepping into stage-3, such as my library: https://github.com/idea2app/MobX-RESTful/blob/de2ec31fa57046ccf4e830bd9d43b085ebfbb779/source/utility/decorator.ts

**Comment by james-pre** [comment]
I went ahead and implemented this in the TS compiler ([james-pre/TypeScript:`json-const-import`](https://github.com/james-pre/TypeScript/tree/json-const-import)), however in order to get merged I'm going to implement it in the Go compiler (i.e. tsgo/v7) as well.

Edit: https://github.com/microsoft/typescript-go/pull/2539 has been opened.

**Comment by james-pre** [comment]
Hey everyone. If you're like me and have work being blocked by this issue you may be interested in a patch for a stable version of TS. I've gone ahead and created one for v5.9.3: https://gist.github.com/james-pre/0764bf9c64428612454d63799dffacc0. You can use something like `patch-package` to apply the feature from #63008 to your install of TS.

Edit:

FYI there was some weirdness with the patch's metadata (file names specifically) since I created it out-of-tree. The new Gist revision should be correct.

`tsc` appears to be working correctly but VS Code's intellisense is very much broken. I'm not sure if this has something to do with the patch or my VS Code configuration.

**Comment by james-pre** [comment]
Okay I made a small mistake when fixing the patch file to work with patch-package. The updated patch in the Gist is working correctly with tsc and with the ts server. Here is a direct link to the patch file: [typescript+5.9.3.patch](https://gist.githubusercontent.com/james-pre/0764bf9c64428612454d63799dffacc0/raw/4c5beee3cd64268412da0ab7ed4df7be3a58536e/typescript+5.9.3.patch)

**Comment by jtannas** [comment]
heyo - just popping in with an interim solution that's as brute-force and reliable as possible
https://www.npmjs.com/package/json2const

```bash
npx json2const --write **/*.json 
```

It generates a `.ts` file for each matched json/jsonc file so you can do a 1-to-1 swap.
```diff
- import data from './data.json';
+ import data from './data.json.ts';
```
I needed something that'd work with top-level JSON arrays so the `.d.ts` solutions weren't working for me.

**Comment by james-pre** [comment]
Hey @jtannas have you tried out my patch? I'd think it would solve your use case really well.

**Comment by jtannas** [comment]
> Hey @jtannas have you tried out my patch? I'd think it would solve your use case really well.

Not yet but I'm interested in trying it.

Unfortunately I saw this issue last week before your patch comments and put together that package. Unlucky timing 🤦

**Comment by jtannas** [comment]
@james-pre 
Heyo - I've been happily using the patch and am looking forward to this being available in an official release. A quick bit of feedback though:

In your PR you state:
> The issue suggests using a `const` import attribute, however this PR uses a compiler option which is much simpler and should be more maintainable.

This is definitely a corner case, but I have mostly normal sized JSON files but also some monster files (>>100k lines). TS does not like this :)
It sees the imported type as `{}` because the JSON is too large & complex for the compiler to reasonably handle. A `const` import attribute can be applied selectively so that I'm not clobbering TS with absurd data files.

**Comment by james-pre** [comment]
Patch for v6.0.2: [typescript+6.0.2.patch](https://gist.github.com/james-pre/0764bf9c64428612454d63799dffacc0/raw/965d61897faee2b5f96be6ce9c5af517c817210b/typescript+6.0.2.patch)

**Comment by james-pre** [comment]
Patch for v6.0.3: [typescript+6.0.3.patch](https://gist.githubusercontent.com/james-pre/0764bf9c64428612454d63799dffacc0/raw/95906a8b477ed78b97425891032f324a510c1750/typescript+6.0.3.patch)

**Comment by Chinoman10** [comment]
It's been 7 years since this feature has been requested 😭 sounds like such an obvious thing to add (case in point, I, much like the first people in 2019, also tried adding `as const` only to realize it wasn't possible, and then found this)...

**Comment by junipertheleprechaun** [comment]
> It's been 7 years since this feature has been requested

Problem is it's (apparently) not a feature microslop needs, and there's no dollar signs attached so...

**Comment by rfon6ngy** [comment]
as long as @james-pre is doing the lords work ❤️, I can survive 
but I agree, it's a pain to patch each updates

**Comment by james-pre** [comment]
npm is introducing built-in dependency patching (`npm patch`) in v12.

You can patch TS by adding this to your package.json:
```json
"patchedDependencies": {
		"typescript@6.0.3": "patches/typescript@6.0.3.patch"
}
```

... and download the corresponding patch (into `patches`), [typescript@6.0.3.patch](https://gist.github.com/james-pre/0764bf9c64428612454d63799dffacc0/raw/aaba969904412ffbe1d950c15c7a679f833fffd6/typescript@6.0.3.patch). The patch will be applied on the next `npm install`.

**Comment by Chinoman10** [comment]
> npm is introducing built-in dependency patching (`npm patch`) in v12.
> 
> You can patch TS by adding this to your package.json:
> 
> "patchedDependencies": {
> 		"typescript@6.0.3": "patches/typescript@6.0.3.patch"
> }
> 
> ... and download the corresponding patch (into `patches`), [typescript@6.0.3.patch](https://gist.github.com/james-pre/0764bf9c64428612454d63799dffacc0/raw/aaba969904412ffbe1d950c15c7a679f833fffd6/typescript@6.0.3.patch). The patch will be applied on the next `npm install`.

Bun has had it for quite a while, and it's much faster than npm as well: https://bun.com/docs/pm/cli/patch
Thanks for the instructions and the good work @james-pre 🙏
